### PR TITLE
Integrate first step

### PR DESCRIPTION
**Pull request by kmilo9999**
- Moved B01_SL_load_single_file.py to src package folder
- Added necessary modules to work with the new config set up
- Fixed variable nomenclature

### CODE REVIEW

File: src/icesat2_tracks/ICEsat2_SI_tools/beam_stats.py [hunk 1]
```diff
@@ -0,0 +1,249 @@
+import numpy as np
+import pandas as pd
+import icesat2_tracks.ICEsat2_SI_tools.spectral_estimates as spec
+#import io as io_local
```
Comment (cpaniaguam) [inline src/icesat2_tracks/ICEsat2_SI_tools/beam_stats.py:4]:
```suggestion
```

File: src/icesat2_tracks/ICEsat2_SI_tools/beam_stats.py [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+
+    # make 2 x 2 plot
+    ax1 = plt.subplot(gs[0, 0])
+    for k in high_beams:
+        plt.plot(D[k]['x']/1e3, np.sqrt(D[k]['var']), '.', color=  col_dict[k], markersize=4, label=k)
+
+    plt.title('high beams std', loc='left')
+    #plt.xlabel('along track distance (km)')
+    plt.ylabel('segment std log(m)')
+    #plt.ylim(0, 20)
```
Comment (cpaniaguam) [inline src/icesat2_tracks/ICEsat2_SI_tools/beam_stats.py:93]:
```suggestion
    plt.ylabel('segment std log(m)')
```

File: src/icesat2_tracks/ICEsat2_SI_tools/beam_stats.py [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+    plt.ylabel('Point Density (m)')
+
+    ax3 = plt.subplot(gs[0, 1])
+    for k in low_beams:
+        plt.plot(D[k]['x']/1e3, np.sqrt(D[k]['var']), '.', color=  col_dict[k], markersize=4, label=k)
+
+    plt.title('low beams std', loc='left')
+    #plt.xlabel('along track distance (km)')
+    #plt.ylabel('segment std (m)')
+    #plt.ylim(0, 20)
```
Comment (cpaniaguam) [inline src/icesat2_tracks/ICEsat2_SI_tools/beam_stats.py:113]:
```suggestion
```

File: src/icesat2_tracks/ICEsat2_SI_tools/beam_stats.py [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+    # plt.xlabel('along track distance (km) or Lat')
+    # #plt.ylabel('Point density (m)')
+    # #return F
+
+    # plt.subplot(2,3, 5)
+    # plt.scatter(D['gt1l']['x']/1e3, D['gt1l']['lat'], s= np.exp(D['gt1l']['N'] *10) , marker='.', color=  col_dict['gt1l'],  label='gt1l', alpha = 0.3)
+
+    # plt.subplots_adjust(left=0.1, bottom=0.1, right=0.9, top=0.9, wspace=0.3, hspace=0.3)
+
+## plot track stats basics for sliderules ATL06 output
```
Comment (cpaniaguam) [inline src/icesat2_tracks/ICEsat2_SI_tools/beam_stats.py:192]:
```suggestion
```

File: src/icesat2_tracks/ICEsat2_SI_tools/sliderule_converter_tools.py [hunk 5]
```diff
@@ -60,12 +74,14 @@ def get_min_eq_dist(ppoly):
     """
     min_eq_dist= list()
     for point in ppoly:
-        point['distance'] = haversine(point['lon'], 0, point['lon'], point['lat'], arc=False)
-        #print(point['lat'], point['distance'])
-        min_eq_dist.append(point['distance'])
+        point['x_atc'] = haversine(point['lon'], 0, point['lon'], point['lat'], arc=False)
+        #print(point['lat'], point['x_atc'])
```
Comment (cpaniaguam) [inline src/icesat2_tracks/ICEsat2_SI_tools/sliderule_converter_tools.py:78]:
```suggestion
```

File: src/icesat2_tracks/ICEsat2_SI_tools/sliderule_converter_tools.py [hunk 6]
```diff
@@ -316,10 +332,10 @@ def define_x_coordinate_in_polygon(table_data, polygon, round=True):
         min_eq_dist = np.round(get_min_eq_dist(polygon))
         
     if ascending_test(table_data):
-        table_data['x'] = table_data['distance'] - min_eq_dist
+        table_data['x'] = table_data['x_atc'] - min_eq_dist
     else:
         #print('descending')
```
Comment (cpaniaguam) [inline src/icesat2_tracks/ICEsat2_SI_tools/sliderule_converter_tools.py:337]:
```suggestion
```

File: src/icesat2_tracks/analysis_db/B01_SL_load_single_file.py [hunk 7]
```diff
@@ -0,0 +1,271 @@
...
+import shapely
+from ipyleaflet import basemaps, Map, GeoData
+
+import icesat2_tracks.ICEsat2_SI_tools.sliderule_converter_tools as sct
+import icesat2_tracks.ICEsat2_SI_tools.io as io
+import icesat2_tracks.ICEsat2_SI_tools.beam_stats as beam_stats
+import icesat2_tracks.local_modules.m_tools_ph3 as MT
+from icesat2_tracks.local_modules import m_general_ph3 as M
+
+#import spicke_remover
```
Comment (cpaniaguam) [inline src/icesat2_tracks/analysis_db/B01_SL_load_single_file.py:23]:
```suggestion
```

File: src/icesat2_tracks/analysis_db/B01_SL_load_single_file.py [hunk 8]
```diff
@@ -0,0 +1,271 @@
...
+
+save_path  = mconfig['paths']['work'] +'/'+batch_key+'/B01_regrid/'
+MT.mkdirs_r(save_path)
+
+save_path_json  = mconfig['paths']['work'] +'/'+ batch_key +'/A01b_ID/'
+MT.mkdirs_r(save_path_json)
+
+#ID, _, _, _ = io.init_data(track_name, batch_key, True, mconfig['paths']['work'],  )
+ATL03_track_name = 'ATL03_'+track_name+'.h5'
+#track_name = ID['tracks']['ATL03'][0] +'.h5'
```
Comment (cpaniaguam) [inline src/icesat2_tracks/analysis_db/B01_SL_load_single_file.py:59]:
```suggestion
# Select region and retrive batch of tracks
track_name, batch_key, ID_flag = io.init_from_input(sys.argv) # loads standard experiment
plot_flag = True
hemis = batch_key.split('_')[0]

save_path  = mconfig['paths']['work'] +'/'+batch_key+'/B01_regrid/'
MT.mkdirs_r(save_path)

save_path_json  = mconfig['paths']['work'] +'/'+ batch_key +'/A01b_ID/'
MT.mkdirs_r(save_path_json)

ATL03_track_name = 'ATL03_'+track_name+'.h5'
```

File: src/icesat2_tracks/analysis_db/B01_SL_load_single_file.py [hunk 9]
```diff
@@ -0,0 +1,271 @@
...
+print("ENDS")
+gdf = sct.correct_and_remove_height(gdf, maximum_height)#.plot(markersize=0.2)
+
+
+# %%
+cdict = dict()
+for s,b in zip(gdf['spot'].unique(), ['gt1l', 'gt1r', 'gt2l', 'gt2r', 'gt3l', 'gt3r']):
+    cdict[s] = color_schemes.rels[b]
+
+#imp.reload(beam_stats)
```
Comment (cpaniaguam) [inline src/icesat2_tracks/analysis_db/B01_SL_load_single_file.py:113]:
```suggestion
```

File: src/icesat2_tracks/analysis_db/B01_SL_load_single_file.py [hunk 10]
```diff
@@ -0,0 +1,271 @@
...
+    inputs:
+        table_data: GeoDataFrame with the data
+        split_by_beam: True/False. If True the data is split by beam
+    returns:
+        if split_by_beam:
+            table_data: dict of GeoDataFrame with the data for each beam
+        else:
+            table_data: GeoDataFrame with the data for all beams in one table
+    """
+    #table_data = copy.copy(B01b_atl06_native)
```
Comment (cpaniaguam) [inline src/icesat2_tracks/analysis_db/B01_SL_load_single_file.py:136]:
```suggestion
```

File: src/icesat2_tracks/analysis_db/B01_SL_load_single_file.py [hunk 11]
```diff
@@ -0,0 +1,271 @@
...
+
+    table_data.rename(columns= {
+                        'n_fit_photons':'N_photos',
+                        'w_surface_window_final':'signal_confidence',
+                        'y_atc':'y',
+                        'x_atc': 'distance',  
+                        #'spot':'beam',
```
Comment (cpaniaguam) [inline src/icesat2_tracks/analysis_db/B01_SL_load_single_file.py:143]:
```suggestion
```

File: src/icesat2_tracks/analysis_db/B01_SL_load_single_file.py [hunk 12]
```diff
@@ -0,0 +1,271 @@
...
+table_data = sct.define_x_coordinate_from_data(table_data)
+table_time = table_data['time']
+table_data.drop(columns=['time'], inplace=True)
+#table_data['time'] = np.datetime_as_string(table_data['time'])#.view('<M8[s]')
+
+# fake 'across'
+#table_data['across'] = table_data['y_atc']*0 +table_data['spot']
+
+#table_data['time'].astype('<S30').view('<M8[s]')
+# add spike remover
```
Comment (cpaniaguam) [inline src/icesat2_tracks/analysis_db/B01_SL_load_single_file.py:180]:
```suggestion
```

File: src/icesat2_tracks/analysis_db/B01_SL_load_single_file.py [hunk 13]
```diff
@@ -0,0 +1,271 @@
...
+Ti = make_B01_dict(table_data, split_by_beam=True, to_hdf5=True) 
+
+for kk in Ti.keys():
+    Ti[kk]['dist'] = Ti[kk]['x'].copy()
+    Ti[kk]['heights_c_weighted_mean'] = Ti[kk]['h_mean'].copy()
+    Ti[kk]['heights_c_std'] = Ti[kk]['h_sigma'].copy()
+
+#Ti[kk]['dist'] = Ti[kk]['x'].copy()
+#Ti['gt1l'].drop('geometry', axis=1, inplace=True)
+
```
Comment (cpaniaguam) [inline src/icesat2_tracks/analysis_db/B01_SL_load_single_file.py:192]:
```suggestion
```

File: src/icesat2_tracks/analysis_db/B01_SL_load_single_file.py [hunk 14]
```diff
@@ -0,0 +1,271 @@
...
+},
+'end': {'longitude': table_data.lons[end_pos], 'latitude': table_data.lats[end_pos]
+, 'seg_dist_x': float(table_data.x[end_pos])
+, 'delta_time': datetime.datetime.timestamp(table_time[end_pos])
+ #table_data.index[end_pos]
+},
+    }
+
+
+#DD['pars']['start']['delta_time'] = str(table_data.index[start_pos])
```
Comment (cpaniaguam) [inline src/icesat2_tracks/analysis_db/B01_SL_load_single_file.py:263]:
```suggestion
},
'end': {'longitude': table_data.lons[end_pos], 'latitude': table_data.lats[end_pos]
, 'seg_dist_x': float(table_data.x[end_pos])
, 'delta_time': datetime.datetime.timestamp(table_time[end_pos])
},
    }
```

File: src/icesat2_tracks/analysis_db/B01_SL_load_single_file.py [hunk 15]
```diff
@@ -0,0 +1,271 @@
...
+
+MT.json_save2(name='A01b_ID_'+ID_name, path=save_path_json, data= DD)
+
+#DD['pars']['start']['delta_time'] = str(table_data.index[start_pos])
+
```
Comment (cpaniaguam) [inline src/icesat2_tracks/analysis_db/B01_SL_load_single_file.py:268]:
```suggestion
```

**Review by cpaniaguam — CHANGES_REQUESTED** [review]
I guess we can start cleaning up the files a bit. If you like, consider hitting them with a formatter (black maybe?). Or even better, add pre-commit hooks? Here is a sample one which might do more than we need right now (type checking and stuff like that.) https://github.com/lnccbrown/HSSM/blob/main/.pre-commit-config.yaml Tha

File: src/icesat2_tracks/ICEsat2_SI_tools/beam_stats.py (hunk 2 above)
Comment (kmilo9999) [reply src/icesat2_tracks/ICEsat2_SI_tools/beam_stats.py]: done

File: src/icesat2_tracks/ICEsat2_SI_tools/sliderule_converter_tools.py [hunk 16]
```diff
@@ -120,16 +135,16 @@ def ascending_test(track_data):
...
+    first_point = abs(track_data.iloc[track_data['x_atc'].argmin()].geometry.y)
+    last_point  = abs(track_data.iloc[track_data['x_atc'].argmax()].geometry.y)
 
-    # first_point = track_data.iloc[track_data.index.argmin()]['distance']
-    # last_point = track_data.iloc[track_data.index.argmax()]['distance']
+    # first_point = track_data.iloc[track_data.index.argmin()]['x_atc']
+    # last_point = track_data.iloc[track_data.index.argmax()]['x_atc']
     # get the time
     # first_time = cGPS.convert_GPS_time(first_point.index, first_point['rgt'], first_point['orbit_number'])
     # last_time = cGPS.convert_GPS_time(last_point['delta_time'], last_point['rgt'], last_point['orbit_number'])
```
Comment (cpaniaguam) [inline src/icesat2_tracks/ICEsat2_SI_tools/sliderule_converter_tools.py:150]: Can these be deleted?
```suggestion
```

File: src/icesat2_tracks/ICEsat2_SI_tools/beam_stats.py (hunk 1 above)
Comment (cpaniaguam) [reply src/icesat2_tracks/ICEsat2_SI_tools/beam_stats.py:4]: Delete this one?

File: src/icesat2_tracks/analysis_db/B01_SL_load_single_file.py [hunk 17]
```diff
@@ -0,0 +1,239 @@
...
+ 'yapc': dict(knn=0, win_h=6, win_x=11, min_ph=4, score=100), # use the YAPC photon classifier; these are the recommended parameters, but the results might be more specific with a smaller win_h value, or a higher score cutoff
+#   "yapc": dict(knn=0, win_h=3, win_x=11, min_ph=4, score=50),  # use the YAPC photon classifier; these are the recommended parameters, but the results might be more specific with a smaller win_h value, or a higher score cutoff
+'atl03_geo_fields' : ['dem_h']
+} 
+
+maximum_height = 30 # (meters) maximum height past dem_h correction 
+print("STARTS")
+gdf = icesat2.atl06p(params_yapc, resources=[ATL03_track_name])
+print("ENDS")
+gdf = sct.correct_and_remove_height(gdf, maximum_height)#.plot(markersize=0.2)
```
Comment (cpaniaguam) [inline src/icesat2_tracks/analysis_db/B01_SL_load_single_file.py:92]:
```suggestion
gdf = sct.correct_and_remove_height(gdf, maximum_height)
```

File: src/icesat2_tracks/analysis_db/B01_SL_load_single_file.py [hunk 18]
```diff
@@ -0,0 +1,239 @@
...
+    table_data['lats'] = table_data['geometry'].y
+
+    drop_columns = ['cycle','gt','rgt', 'pflags']
+    if to_hdf5:
+        drop_columns.append('geometry')
+    table_data.drop(columns=drop_columns, inplace=True)
+
+    if split_by_beam:
+        B01b = dict()
+        # this is not tested
```
Comment (cpaniaguam) [inline src/icesat2_tracks/analysis_db/B01_SL_load_single_file.py:141]: Seeing comments like this one is a bit unsettling to me. Do we know if this has been tested?

File: src/icesat2_tracks/analysis_db/B01_SL_load_single_file.py [hunk 19]
```diff
@@ -0,0 +1,239 @@
...
+    F.save_light(path = plot_path , name = 'B01b_beam_statistics.png')
+    plt.close()
+
+    # plot the ground tracks in geographic location
+    gdf[::100].plot(markersize=0.1, figsize=(4,6))
+    plt.title(track_name +  '\nascending =' + str(sct.ascending_test_distance(gdf)) , loc ='left')
+    # gdf_atl03 = icesat2.atl03s(params, [ATL03_track_name])
+    M.save_anyfig(plt.gcf(), path = plot_path  ,name = 'B01_track.png')
+    # gdf_atl03.plot()
+    plt.close()
```
Comment (cpaniaguam) [inline src/icesat2_tracks/analysis_db/B01_SL_load_single_file.py:205]:
```suggestion
    plt.title(track_name +  '\nascending =' + str(sct.ascending_test_distance(gdf)) , loc ='left')
    M.save_anyfig(plt.gcf(), path = plot_path  ,name = 'B01_track.png')
    plt.close()
```

**Review by cpaniaguam — CHANGES_REQUESTED** [review]
Couple more cleanups suggested..

File: src/icesat2_tracks/ICEsat2_SI_tools/sliderule_converter_tools.py (hunk 16 above)
Comment (kmilo9999) [reply src/icesat2_tracks/ICEsat2_SI_tools/sliderule_converter_tools.py:150]: @mochell what you think?

File: src/icesat2_tracks/analysis_db/B01_SL_load_single_file.py (hunk 18 above)
Comment (kmilo9999) [reply src/icesat2_tracks/analysis_db/B01_SL_load_single_file.py:141]: It works on my side. @mochell any thoughts ?

Comment (hollandjg) [reply src/icesat2_tracks/analysis_db/B01_SL_load_single_file.py:141]: I think we need to live with the unsettling feeling. This should be fine.

File: src/icesat2_tracks/ICEsat2_SI_tools/sliderule_converter_tools.py [hunk 20]
```diff
@@ -120,16 +135,16 @@ def ascending_test(track_data):
 
 def ascending_test_distance(track_data):
     """
-    test if the track is ascending or descending based on 'distance' column
+    test if the track is ascending or descending based on 'x_atc' column
```
[the remark this answers is not in the document]
Comment (hollandjg) [reply src/icesat2_tracks/ICEsat2_SI_tools/sliderule_converter_tools.py:138]: This is the column name we get from sliderule.

File: src/icesat2_tracks/ICEsat2_SI_tools/sliderule_converter_tools.py (hunk 16 above)
Comment (kmilo9999) [reply src/icesat2_tracks/ICEsat2_SI_tools/sliderule_converter_tools.py]: Commented code removed

File: src/icesat2_tracks/analysis_db/B01_SL_load_single_file.py [hunk 21]
```diff
@@ -0,0 +1,239 @@
...
+        B01b = dict()
+        # this is not tested
+        for spot,beam in zip([1,2,3, 4, 5, 6],['gt1l', 'gt1r', 'gt2l', 'gt2r', 'gt3l', 'gt3r']) : 
+            ii = table_data.spot==spot
+            B01b[beam] = table_data[ii]
+        return B01b
+    else:
+        return table_data
+
+# %%
```
[the remark this answers is not in the document]
Comment (kmilo9999) [reply src/icesat2_tracks/analysis_db/B01_SL_load_single_file.py]: removed these marks from this specific file.

**Review by hollandjg — APPROVED** [review]
Rocker🚀

**Pull request merged**